File: frontend/apps/ios/Pantopus/Core/Analytics/Analytics.swift
//
//  Analytics.swift
//  Pantopus
//
//  Typed analytics taxonomy + shim. The taxonomy is closed — any new
//  event must be added to `AnalyticsEvent` so untyped strings never
//  reach the wire. In Debug builds events log to the console; in
//  Release they currently no-op (vendor SDK lands later).
//
//  TODO(analytics): wire `track(_:)` to the real vendor (Amplitude /
//  Mixpanel / PostHog). Today we only mirror events into the Sentry
//  breadcrumb stream via `Observability.shared.track`.
//

import Foundation

/// Closed list of analytics events. Mirrors the P15 taxonomy 1:1; do
/// not invent ad-hoc names — extend the enum instead.
public enum AnalyticsEvent: Sendable, Equatable {
    case screenHubViewed
    case screenMailboxListViewed
    case screenMailboxDrawersViewed
    case screenMyHomesViewed
    case screenHomeDashboardViewed
    case screenMailboxItemDetailViewed(category: String, trustLevel: String)
    case screenEditProfileViewed
    case screenAddHomeWizardStepViewed(stepNumber: Int, stepName: String)
    case screenClaimOwnershipStepViewed(stepName: String)
    case screenMyClaimsViewed
    case screenBillsViewed
    case screenBillDetailViewed
    case screenAddBillWizardStepViewed(stepNumber: Int, stepName: String)
    case ctaAddBillSubmit(result: AnalyticsResult)
    case screenHomeMaintenanceViewed
    case screenLogMaintenanceViewed
    case screenMaintenanceDetailViewed
    case ctaLogMaintenanceSubmit(result: AnalyticsResult)
    case ctaMaintenanceDelete(result: AnalyticsResult)
    case screenPetsListViewed
    case screenPetsWizardStepViewed(stepNumber: Int, stepName: String)
    case screenHomeCalendarViewed
    case screenEmergencyInfoViewed
    case screenDocumentsViewed
    case screenPackagesViewed
    case screenPackageDetailViewed
    case ctaLogPackageSubmit(result: AnalyticsResult)
    case screenPollsViewed
    case screenPollDetailViewed
    case ctaPollVoteSubmit(result: AnalyticsResult)
    case screenHouseholdTasksViewed
    case screenOwnersListViewed
    case screenMembersListViewed
    case screenMembersWizardStepViewed(stepNumber: Int, stepName: String)
    case screenPulseFeedViewed(intent: String)
    /// T6.3f / P14 — My listings index (the seller's tabbed list).
    case screenMyListingsViewed
    /// T6.3f / P14 — My businesses index (owner/staff roster).
    case screenMyBusinessesViewed
    case ctaHubActionStripTapped(label: String)
    case ctaHubPillarTapped(pillar: String)
    case ctaMailboxItemLogReceived
    case ctaAddHomeSubmit
    /// P2.3 — Snap & Sell wizard step view event.
    case screenListingComposeWizardStepViewed(stepNumber: Int, stepName: String)
    /// P2.3 — submit the listing-compose wizard (final POST).
    case ctaListingComposeSubmit
    case ctaClaimOwnershipSubmit(result: AnalyticsResult)
    case formEditProfileSubmit(result: AnalyticsResult)
    case formEditProfileValidationError(field: String)
    /// P2.1 — Pulse compose form submitted. `intent` is one of
    /// `ask / recommend / event / lost / announce`.
    case formPulseComposeSubmit(intent: String, result: AnalyticsResult)
    /// P2.1 — Pulse compose validation failure. `intent` matches the
    /// form's active variant; `field` is the failing field's id.
    case formPulseComposeValidationError(intent: String, field: String)
    /// P2.1 — Pulse compose screen viewed.
    case screenPulseComposeViewed(intent: String)
    /// P2.2 — Post-a-Task wizard step view.
    case screenComposeGigWizardStepViewed(stepNumber: Int, stepName: String)
    /// P2.2 — Post-a-Task wizard submit tap (fires before the POST).
    case ctaComposeGigSubmit

    /// Wire-format event name. Stable across versions — vendor / dashboard
    /// owners depend on these strings.
    public var name: String {
        switch self {
        case .screenHubViewed: "screen.hub.viewed"
        case .screenMailboxListViewed: "screen.mailbox_list.viewed"
        case .screenMailboxDrawersViewed: "screen.mailbox_drawers.viewed"
        case .screenMyHomesViewed: "screen.my_homes.viewed"
        case .screenHomeDashboardViewed: "screen.home_dashboard.viewed"
        case .screenMailboxItemDetailViewed: "screen.mailbox_item_detail.viewed"
        case .screenEditProfileViewed: "screen.edit_profile.viewed"
        case .screenAddHomeWizardStepViewed: "screen.add_home_wizard.step_viewed"
        case .screenClaimOwnershipStepViewed: "screen.claim_ownership_wizard.step_viewed"
        case .screenMyClaimsViewed: "screen.my_claims.viewed"
        case .screenBillsViewed: "screen.bills.viewed"
        case .screenBillDetailViewed: "screen.bill_detail.viewed"
        case .screenAddBillWizardStepViewed: "screen.add_bill_wizard.step_viewed"
        case .ctaAddBillSubmit: "cta.add_bill.submit"
        case .screenHomeMaintenanceViewed: "screen.home_maintenance.viewed"
        case .screenLogMaintenanceViewed: "screen.log_maintenance.viewed"
        case .screenMaintenanceDetailViewed: "screen.maintenance_detail.viewed"
        case .ctaLogMaintenanceSubmit: "cta.log_maintenance.submit"
        case .ctaMaintenanceDelete: "cta.maintenance.delete"
        case .screenPetsListViewed: "screen.pets_list.viewed"
        case .screenPetsWizardStepViewed: "screen.pets_wizard.step_viewed"
        case .screenHomeCalendarViewed: "screen.home_calendar.viewed"
        case .screenEmergencyInfoViewed: "screen.emergency_info.viewed"
        case .screenDocumentsViewed: "screen.documents.viewed"
        case .screenPackagesViewed: "screen.packages.viewed"
        case .screenPackageDetailViewed: "screen.package_detail.viewed"
        case .ctaLogPackageSubmit: "cta.log_package.submit"
        case .screenPollsViewed: "screen.polls.viewed"
        case .screenPollDetailViewed: "screen.poll_detail.viewed"
        case .ctaPollVoteSubmit: "cta.poll_vote.submit"
        case .screenHouseholdTasksViewed: "screen.household_tasks.viewed"
        case .screenOwnersListViewed: "screen.owners_list.viewed"
        case .screenMembersListViewed: "screen.members_list.viewed"
        case .screenMembersWizardStepViewed: "screen.members_wizard.step_viewed"
        case .screenPulseFeedViewed: "screen.pulse_feed.viewed"
        case .screenMyListingsViewed: "screen.my_listings.viewed"
        case .screenMyBusinessesViewed: "screen.my_businesses.viewed"
        case .ctaHubActionStripTapped: "cta.hub.action_strip_tapped"
        case .ctaHubPillarTapped: "cta.hub.pillar_tapped"
        case .ctaMailboxItemLogReceived: "cta.mailbox_item.log_received"
        case .ctaAddHomeSubmit: "cta.add_home.submit"
        case .screenListingComposeWizardStepViewed: "screen.listing_compose_wizard.step_viewed"
        case .ctaListingComposeSubmit: "cta.listing_compose.submit"
        case .ctaClaimOwnershipSubmit: "cta.claim_ownership.submit"
        case .formEditProfileSubmit: "form.edit_profile.submit"
        case .formEditProfileValidationError: "form.edit_profile.validation_error"
        case .formPulseComposeSubmit: "form.pulse_compose.submit"
        case .formPulseComposeValidationError: "form.pulse_compose.validation_error"
        case .screenPulseComposeViewed: "screen.pulse_compose.viewed"
        case .screenComposeGigWizardStepViewed: "screen.compose_gig_wizard.step_viewed"
        case .ctaComposeGigSubmit: "cta.compose_gig.submit"
        }
    }

    /// Per-event property bag. Properties are flat string-keyed so the
    /// vendor schema doesn't need nested types.
    public var properties: [String: String] {
        switch self {
        case let .screenMailboxItemDetailViewed(category, trustLevel):
            ["category": category, "trust_level": trustLevel]
        case let .screenAddHomeWizardStepViewed(stepNumber, stepName):
            ["step_number": "\(stepNumber)", "step_name": stepName]
        case let .screenComposeGigWizardStepViewed(stepNumber, stepName):
            ["step_number": "\(stepNumber)", "step_name": stepName]
        case let .screenListingComposeWizardStepViewed(stepNumber, stepName):
            ["step_number": "\(stepNumber)", "step_name": stepName]
        case let .screenPetsWizardStepViewed(stepNumber, stepName):
            ["step_number": "\(stepNumber)", "step_name": stepName]
        case let .screenMembersWizardStepViewed(stepNumber, stepName):
            ["step_number": "\(stepNumber)", "step_name": stepName]
        case let .screenClaimOwnershipStepViewed(stepName):
            ["step_name": stepName]
        case let .ctaHubActionStripTapped(label):
            ["label": label]
        case let .ctaHubPillarTapped(pillar):
            ["pillar": pillar]
        case let .screenPulseFeedViewed(intent):
            ["intent": intent]
        case let .ctaClaimOwnershipSubmit(result):
            ["result": result.rawValue]
        case let .formEditProfileSubmit(result):
            ["result": result.rawValue]
        case let .formEditProfileValidationError(field):
            ["field": field]
        case let .formPulseComposeSubmit(intent, result):
            ["intent": intent, "result": result.rawValue]
        case let .formPulseComposeValidationError(intent, field):
            ["intent": intent, "field": field]
        case let .screenPulseComposeViewed(intent):
            ["intent": intent]
        case let .screenAddBillWizardStepViewed(stepNumber, stepName):
            ["step_number": "\(stepNumber)", "step_name": stepName]
        case let .ctaAddBillSubmit(result):
            ["result": result.rawValue]
        case let .ctaLogPackageSubmit(result):
            ["result": result.rawValue]
        case let .ctaLogMaintenanceSubmit(result):
            ["result": result.rawValue]
        case let .ctaMaintenanceDelete(result):
            ["result": result.rawValue]
        case let .ctaPollVoteSubmit(result):
            ["result": result.rawValue]
        case .screenHubViewed,
             .screenMailboxListViewed,
             .screenMailboxDrawersViewed,
             .screenMyHomesViewed,
             .screenHomeDashboardViewed,
             .screenMyClaimsViewed,
             .screenBillsViewed,
             .screenBillDetailViewed,
             .screenHomeMaintenanceViewed,
             .screenLogMaintenanceViewed,
             .screenMaintenanceDetailViewed,
             .screenPetsListViewed,
             .screenHomeCalendarViewed,
             .screenPackagesViewed,
             .screenPackageDetailViewed,
             .screenPollsViewed,
             .screenPollDetailViewed,
             .screenMyListingsViewed,
             .screenMyBusinessesViewed,
             .screenHouseholdTasksViewed,
             .screenOwnersListViewed,
             .screenMembersListViewed,
             .screenEmergencyInfoViewed,
             .screenDocumentsViewed,
             .screenEditProfileViewed,
             .ctaMailboxItemLogReceived,
             .ctaAddHomeSubmit,
             .ctaComposeGigSubmit,
             .ctaListingComposeSubmit:
            [:]
        }
    }
}

/// Standard outcomes for form submissions and other yes/no telemetry.
public enum AnalyticsResult: String, Sendable {
    case success
    case error
}

/// Analytics shim. Feature code calls `Analytics.track(.screenHubViewed)`
/// and we route to the Observability layer. In Debug builds the event
/// also lands in the Xcode console for quick verification.
public enum Analytics {
    public static func track(_ event: AnalyticsEvent) {
        #if DEBUG
        let propsDescription = event.properties.isEmpty
            ? ""
            : " " + event.properties.map { "\($0.key)=\($0.value)" }.joined(separator: " ")
        print("📊 analytics \(event.name)\(propsDescription)")
        #endif
        Task { @MainActor in
            Observability.shared.track(event.name, properties: event.properties)
        }
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/analytics/Analytics.kt
package app.pantopus.android.data.analytics

import app.pantopus.android.BuildConfig
import app.pantopus.android.data.observability.Observability
import timber.log.Timber

/**
 * Closed list of analytics events. Mirrors the P15 taxonomy 1:1; do
 * not invent ad-hoc names — extend the sealed hierarchy instead.
 */
sealed class AnalyticsEvent(
    val eventName: String,
) {
    open val properties: Map<String, String> get() = emptyMap()

    data object ScreenHubViewed : AnalyticsEvent("screen.hub.viewed")

    data object ScreenMailboxListViewed : AnalyticsEvent("screen.mailbox_list.viewed")

    data object ScreenMailboxDrawersViewed : AnalyticsEvent("screen.mailbox_drawers.viewed")

    data object ScreenMyHomesViewed : AnalyticsEvent("screen.my_homes.viewed")

    data object ScreenHomeDashboardViewed : AnalyticsEvent("screen.home_dashboard.viewed")

    data object ScreenEditProfileViewed : AnalyticsEvent("screen.edit_profile.viewed")

    data object ScreenMyClaimsViewed : AnalyticsEvent("screen.my_claims.viewed")

    data object ScreenBillsViewed : AnalyticsEvent("screen.bills.viewed")

    data object ScreenBillDetailViewed : AnalyticsEvent("screen.bill_detail.viewed")

    data class ScreenAddBillWizardStepViewed(
        val stepNumber: Int,
        val stepName: String,
    ) : AnalyticsEvent("screen.add_bill_wizard.step_viewed") {
        override val properties =
            mapOf("step_number" to stepNumber.toString(), "step_name" to stepName)
    }

    data class CtaAddBillSubmit(
        val result: AnalyticsResult,
    ) : AnalyticsEvent("cta.add_bill.submit") {
        override val properties = mapOf("result" to result.value)
    }

    /** T6.3b / P10 — Maintenance list view. */
    data object ScreenHomeMaintenanceViewed : AnalyticsEvent("screen.home_maintenance.viewed")

    data object ScreenNotificationsViewed : AnalyticsEvent("screen.notifications.viewed")

    data object ScreenHomeCalendarViewed : AnalyticsEvent("screen.home_calendar.viewed")

    data object ScreenPetsListViewed : AnalyticsEvent("screen.pets_list.viewed")

    data object ScreenEmergencyInfoViewed : AnalyticsEvent("screen.emergency_info.viewed")

    data object ScreenDocumentsViewed : AnalyticsEvent("screen.documents.viewed")

    data object ScreenPollsViewed : AnalyticsEvent("screen.polls.viewed")

    data object ScreenPollDetailViewed : AnalyticsEvent("screen.poll_detail.viewed")

    data class CtaPollVoteSubmit(
        val result: AnalyticsResult,
    ) : AnalyticsEvent("cta.poll_vote.submit") {
        override val properties = mapOf("result" to result.value)
    }

    /** T6.3f / P14 — My listings index (the seller's tabbed list). */
    data object ScreenMyListingsViewed : AnalyticsEvent("screen.my_listings.viewed")

    /** T6.3f / P14 — My businesses index (owner / staff roster). */
    data object ScreenMyBusinessesViewed : AnalyticsEvent("screen.my_businesses.viewed")

    data object ScreenHouseholdTasksViewed : AnalyticsEvent("screen.household_tasks.viewed")

    data object ScreenOwnersListViewed : AnalyticsEvent("screen.owners_list.viewed")

    data class ScreenPetsWizardStepViewed(
        val stepNumber: Int,
        val stepName: String,
    ) : AnalyticsEvent("screen.pets_wizard.step_viewed") {
        override val properties =
            mapOf("step_number" to stepNumber.toString(), "step_name" to stepName)
    }

    data object ScreenPackagesViewed : AnalyticsEvent("screen.packages.viewed")

    data object ScreenPackageDetailViewed : AnalyticsEvent("screen.package_detail.viewed")

    data class CtaLogPackageSubmit(
        val result: AnalyticsResult,
    ) : AnalyticsEvent("cta.log_package.submit") {
        override val properties = mapOf("result" to result.value)
    }

    data object ScreenMembersListViewed : AnalyticsEvent("screen.members_list.viewed")

    data class ScreenMembersWizardStepViewed(
        val stepNumber: Int,
        val stepName: String,
    ) : AnalyticsEvent("screen.members_wizard.step_viewed") {
        override val properties =
            mapOf("step_number" to stepNumber.toString(), "step_name" to stepName)
    }

    data class ScreenPulseFeedViewed(
        val intent: String,
    ) : AnalyticsEvent("screen.pulse_feed.viewed") {
        override val properties = mapOf("intent" to intent)
    }

    data object CtaMailboxItemLogReceived : AnalyticsEvent("cta.mailbox_item.log_received")

    data object CtaAddHomeSubmit : AnalyticsEvent("cta.add_home.submit")

    /** P2.3 — Snap & Sell wizard step view event. */
    data class ScreenListingComposeWizardStepViewed(
        val stepNumber: Int,
        val stepName: String,
    ) : AnalyticsEvent("screen.listing_compose_wizard.step_viewed") {
        override val properties =
            mapOf("step_number" to stepNumber.toString(), "step_name" to stepName)
    }

    /** P2.3 — submit the listing-compose wizard (final POST). */
    data object CtaListingComposeSubmit : AnalyticsEvent("cta.listing_compose.submit")

    data class ScreenClaimOwnershipStepViewed(
        val stepName: String,
    ) : AnalyticsEvent("screen.claim_ownership_wizard.step_viewed") {
        override val properties = mapOf("step_name" to stepName)
    }

    data class CtaClaimOwnershipSubmit(
        val result: AnalyticsResult,
    ) : AnalyticsEvent("cta.claim_ownership.submit") {
        override val properties = mapOf("result" to result.value)
    }

    data class ScreenMailboxItemDetailViewed(
        val category: String,
        val trustLevel: String,
    ) : AnalyticsEvent("screen.mailbox_item_detail.viewed") {
        override val properties = mapOf("category" to category, "trust_level" to trustLevel)
    }

    data class ScreenAddHomeWizardStepViewed(
        val stepNumber: Int,
        val stepName: String,
    ) : AnalyticsEvent("screen.add_home_wizard.step_viewed") {
        override val properties =
            mapOf("step_number" to stepNumber.toString(), "step_name" to stepName)
    }

    data class CtaHubActionStripTapped(
        val label: String,
    ) : AnalyticsEvent("cta.hub.action_strip_tapped") {
        override val properties = mapOf("label" to label)
    }

    data class CtaHubPillarTapped(
        val pillar: String,
    ) : AnalyticsEvent("cta.hub.pillar_tapped") {
        override val properties = mapOf("pillar" to pillar)
    }

    data class FormEditProfileSubmit(
        val result: AnalyticsResult,
    ) : AnalyticsEvent("form.edit_profile.submit") {
        override val properties = mapOf("result" to result.value)
    }

    data class FormEditProfileValidationError(
        val field: String,
    ) : AnalyticsEvent("form.edit_profile.validation_error") {
        override val properties = mapOf("field" to field)
    }

    /** P2.1 — Pulse compose screen viewed. */
    data class ScreenPulseComposeViewed(
        val intent: String,
    ) : AnalyticsEvent("screen.pulse_compose.viewed") {
        override val properties = mapOf("intent" to intent)
    }

    /** P2.1 — Pulse compose form submitted. */
    data class FormPulseComposeSubmit(
        val intent: String,
        val result: AnalyticsResult,
    ) : AnalyticsEvent("form.pulse_compose.submit") {
        override val properties = mapOf("intent" to intent, "result" to result.value)
    }

    /** P2.1 — Pulse compose validation failed before submit. */
    data class FormPulseComposeValidationError(
        val intent: String,
        val field: String,
    ) : AnalyticsEvent("form.pulse_compose.validation_error") {
        override val properties = mapOf("intent" to intent, "field" to field)
    }

    /** P2.2 — Post-a-Task wizard step view. */
    data class ScreenComposeGigWizardStepViewed(
        val stepNumber: Int,
        val stepName: String,
    ) : AnalyticsEvent("screen.compose_gig_wizard.step_viewed") {
        override val properties =
            mapOf("step_number" to stepNumber.toString(), "step_name" to stepName)
    }

    /** P2.2 — Post-a-Task wizard submit tap (fires before the POST). */
    data object CtaComposeGigSubmit : AnalyticsEvent("cta.compose_gig.submit")
}

/** Standard outcomes for form submissions and other yes/no telemetry. */
enum class AnalyticsResult(
    val value: String,
) {
    SUCCESS("success"),
    ERROR("error"),
}

/**
 * Analytics shim. Feature code calls
 * `Analytics.track(AnalyticsEvent.ScreenHubViewed)` and we route to
 * the [Observability] layer (which mirrors into Sentry breadcrumbs).
 *
 * In Debug builds the event also lands in `Timber` so it shows up in
 * Logcat for quick verification.
 *
 * TODO(analytics): wire this to a real analytics vendor (Amplitude /
 * Mixpanel / PostHog) — today we only mirror events into Sentry.
 */
object Analytics {
    @Volatile private var observability: Observability? = null

    /**
     * Wire the singleton [Observability] instance into the shim. Call
     * once from `PantopusApplication.onCreate`. Tests can re-bind via
     * [bindForTesting].
     */
    fun bind(observability: Observability) {
        this.observability = observability
    }

    /** For test harnesses to swap the bound observability. */
    @Suppress("unused")
    fun bindForTesting(observability: Observability?) {
        this.observability = observability
    }

    fun track(event: AnalyticsEvent) {
        if (BuildConfig.DEBUG) {
            val props =
                if (event.properties.isEmpty()) {
                    ""
                } else {
                    " " + event.properties.entries.joinToString(" ") { "${it.key}=${it.value}" }
                }
            Timber.tag("analytics").i("📊 ${event.eventName}$props")
        }
        observability?.track(event.eventName, event.properties)
    }
}
